hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -3971,11 +3972,139 @@ func (r *HostedClusterReconciler) validateHostedClusterSupport(hc *hyperv1.Hoste
+	}
+
+	if hc.Spec.Networking.APIServer != nil && hc.Spec.Networking.APIServer.AdvertiseAddress != nil {
+		networks["spec.networking.APIServerNetworking.AdvertiseAddress"] = *hc.Spec.Networking.APIServer.AdvertiseAddress
+	} else {
+		// relaxing err checking when we use default AdvertiseAddress
+		ipv4, _ := hyperutil.IsIPv4(hc.Spec.Networking.ClusterNetwork[0].CIDR.String())
+
+		if !ipv4 {
+			networks["spec.networking.APIServerNetworking.AdvertiseAddress"] = config.DefaultAdvertiseIPv6Address
+		} else {
+			networks["spec.networking.APIServerNetworking.AdvertiseAddress"] = config.DefaultAdvertiseIPv4Address
+		}
+	}
+
+	for fieldpath, ipaddr := range networks {
+		checkIP := net.ParseIP(ipaddr)
+
+		if checkIP != nil && strings.Contains(ipaddr, ".") {
+			ipv4 = append(ipv4, ipaddr)
+		}
+
+		if checkIP != nil && strings.Contains(ipaddr, ":") {
+			ipv6 = append(ipv6, ipaddr)
+		}
+
+		// This check ensures that the IPv6 and IPv4 is a valid ip
+		if checkIP == nil {
+			errs = append(errs, field.Invalid(field.NewPath(fieldpath),
+				k8sutilspointer.String(networks[fieldpath]),
+				fmt.Sprintf("error checking network stack of %s with ip %s", fieldpath, ipaddr),
+			))
+		}
+	}
+
+	if len(ipv4) > 0 && len(ipv6) > 0 {
+		// Invalid result, means that there are mixed stacks in the primary position of the stack
+		errs = append(errs, field.Forbidden(field.NewPath("spec.networking"),
+			fmt.Sprintf("declare multiple network stacks as primary network in the cluster definition is not allowed, ipv4: %v, ipv6: %v", ipv4, ipv6),
+		))
+	}
+
+	return errs
+
+}
+
+// checkAdvertiseAddressOverlapping validates that the AdvertiseAddress defined does not overlap with
+// the ClusterNetwork, ServiceNetwork and MachineNetwork
+func checkAdvertiseAddressOverlapping(hc *hyperv1.HostedCluster) field.ErrorList {
+	var errs field.ErrorList
+	var err error
+	var advAddress netip.Addr
+
+	advAddrPath := "spec.networking.APIServerNetworking.AdvertiseAddress"
+	networks := map[string]string{
+		"spec.networking.ServiceNetwork": hc.Spec.Networking.ServiceNetwork[0].CIDR.String(),
+		"spec.networking.MachineNetwork": hc.Spec.Networking.MachineNetwork[0].CIDR.String(),
+		"spec.networking.ClusterNetwork": hc.Spec.Networking.ClusterNetwork[0].CIDR.String(),
+	}
+
+	if hc.Spec.Networking.APIServer != nil && hc.Spec.Networking.APIServer.AdvertiseAddress != nil {
+		advAddress, err = netip.ParseAddr(*hc.Spec.Networking.APIServer.AdvertiseAddress)
+		if err != nil {
+			errs = append(errs, field.Invalid(field.NewPath(advAddrPath),
+				*hc.Spec.Networking.APIServer.AdvertiseAddress,
+				fmt.Sprintf("error parsing AdvertiseAddress: %v", err),
+			))
+		}
+	} else {
+		primaryStackIpv4, err := hyperutil.IsIPv4(networks["spec.networking.ClusterNetwork"])
+		if err != nil {
+			errs = append(errs, field.Invalid(field.NewPath(advAddrPath),
+				k8sutilspointer.String(networks["spec.networking.ClusterNetwork"]),
+				fmt.Sprintf("error checking ClusterNetwork stack: %v", err),
```

Comment:
you probably want to render the error string so this would be %s, rather than %v